### PR TITLE
Gh-6-QR-Code-Scanner-Screen

### PR DESCRIPTION
**Pull request by MouazAlzahabi**

### CODE REVIEW

File: Multisig/Presentation/UI Library/QRScanner/QRCodeScannerViewController.swift [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+
+class QRCodeScannerViewController: UIViewController {
+    @IBOutlet weak var headerLabel: UILabel!
+    var previewLayer: AVCaptureVideoPreviewLayer!
+    var captureSession: AVCaptureSession!
+    var delegate: QRCodeScannerViewControllerDelegate?
+    var header: String?
+    
+    enum Strings {
+        static let cameraAlertTitle = NSLocalizedString("camera_title", comment: "")
```
Comment (sche) [inline Multisig/Presentation/UI Library/QRScanner/QRCodeScannerViewController.swift:24]: My suggestion would be to use a localization key value in English as a comment, or to use some short description if localization key value is too long.

File: Multisig/Presentation/UI Library/QRScanner/QRCodeScannerViewController.swift [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+    }
+    
+    func setup() {
+        AVCaptureDevice.requestAccess(for: .video) { [weak self] (result)  in
+            if result {
+                DispatchQueue.main.async {
+                    self?.createScanner()
+                }
+            }
+            else {
```
Comment (sche) [inline Multisig/Presentation/UI Library/QRScanner/QRCodeScannerViewController.swift:65]: we did not discuss coding style before, but I would suggest for the consistency to put 'else' near the closing bracket of 'if' as you did it below in this file

File: Multisig/Presentation/UI Library/QRScanner/QRCodeScannerViewController.swift [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+    func createScanner() {
+        captureSession = AVCaptureSession()
+
+        guard let videoCaptureDevice = AVCaptureDevice.default(for: .video) else { return }
+        let videoInput: AVCaptureDeviceInput
+
+        do {
+            videoInput = try AVCaptureDeviceInput(device: videoCaptureDevice)
+        } catch {
+            return presentFaildToCreateScannerAlert()
```
Comment (sche) [inline Multisig/Presentation/UI Library/QRScanner/QRCodeScannerViewController.swift:82]: please fix the type: 'failed '

File: Multisig/Presentation/UI Library/QRScanner/QRCodeScannerViewController.swift [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+
+    override var preferredStatusBarStyle: UIStatusBarStyle {
+        return .lightContent
+    }
+}
+
+extension QRCodeScannerViewController: AVCaptureMetadataOutputObjectsDelegate {
+    func metadataOutput(_ output: AVCaptureMetadataOutput, didOutput metadataObjects: [AVMetadataObject], from connection: AVCaptureConnection) {
+        let metadataObject = metadataObjects.first { (metaObject) -> Bool in
+            guard let readableObject = metaObject as? AVMetadataMachineReadableCodeObject else { return false}
```
Comment (sche) [inline Multisig/Presentation/UI Library/QRScanner/QRCodeScannerViewController.swift:149]: please add a space after 'false'

**Comment by sche** [comment]
good job 👍

File: Multisig/Presentation/UI Library/QRScanner/QRCodeScanner.swift [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+//  Created by Moaaz on 4/15/20.
+//  Copyright © 2020 Gnosis Ltd. All rights reserved.
+//
+
+import SwiftUI
+
+struct QRCodeScanner: UIViewControllerRepresentable {
+    var header: String?
+    var handler: (String) -> Void
+    func makeCoordinator() -> Coordinator {
```
Comment (DmitryBespalov) [inline Multisig/Presentation/UI Library/QRScanner/QRCodeScanner.swift:14]: space before function is missing

File: Multisig/Presentation/UI Library/QRScanner/QRCodeScanner.swift [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+        let scannerViewController = QRCodeScannerViewController(nibName: "QRCodeScannerViewController", bundle: nil)
+        scannerViewController.delegate = context.coordinator
+        scannerViewController.header = header
+        scannerViewController.setup()
+        
+        return scannerViewController
+    }
+    
+    func updateUIViewController(_ vc: QRCodeScannerViewController, context: Context) {
+    }
```
Comment (DmitryBespalov) [inline Multisig/Presentation/UI Library/QRScanner/QRCodeScanner.swift:28]: please add a comment to the empty body
```suggestion
        // do nothing
    }
```

File: Multisig/Presentation/UI Library/QRScanner/QRCodeScannerViewController.swift [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        headerLabel.text = header
+    }
+
+    override func viewWillAppear(_ animated: Bool) {
+        super.viewWillAppear(animated)
+
+        if (captureSession?.isRunning == false) {
```
Comment (DmitryBespalov) [inline Multisig/Presentation/UI Library/QRScanner/QRCodeScannerViewController.swift:41]: are parentheses here required? I think no, and parentheses-less is a preferred coding style. (Also please check other conditional statements below that have the same issue)

File: Multisig/Presentation/UI Library/QRScanner/QRCodeScannerViewController.swift [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+                    self?.presentCameraAccessRequiredAlert()
+                }
+            }
+        }
+    }
+
+    func createScanner() {
+        captureSession = AVCaptureSession()
+
+        guard let videoCaptureDevice = AVCaptureDevice.default(for: .video) else { return }
```
Comment (DmitryBespalov) [inline Multisig/Presentation/UI Library/QRScanner/QRCodeScannerViewController.swift:76]: should this be before creating capture session?

File: Multisig/Presentation/UI Library/QRScanner/QRCodeScannerViewController.swift [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+extension QRCodeScannerViewController: AVCaptureMetadataOutputObjectsDelegate {
+    func metadataOutput(_ output: AVCaptureMetadataOutput, didOutput metadataObjects: [AVMetadataObject], from connection: AVCaptureConnection) {
+        let metadataObject = metadataObjects.first { (metaObject) -> Bool in
+            guard let readableObject = metaObject as? AVMetadataMachineReadableCodeObject else { return false}
+            return !(readableObject.stringValue?.isEmpty ?? false) && readableObject.type == .qr
+        } as? AVMetadataMachineReadableCodeObject
+        
+        if let metadataObject = metadataObject {
+            AudioServicesPlaySystemSound(SystemSoundID(kSystemSoundID_Vibrate))
+            scannerDidScan(code: metadataObject.stringValue!)
```
Comment (DmitryBespalov) [inline Multisig/Presentation/UI Library/QRScanner/QRCodeScannerViewController.swift:155]: I think rewriting this method will be more readable:
```suggestion
    func metadataOutput(_ output: AVCaptureMetadataOutput,
                        didOutput metadataObjects: [AVMetadataObject],
                        from connection: AVCaptureConnection) {
        let scannedValue = metadataObjects
            .compactMap { $0 as? AVMetadataMachineReadableCodeObject }
            .filter { $0.type == .qr }
            .compactMap { $0.stringValue }
            .filter { !$0.isEmpty }
            .first
        
        if let code = scannedValue {
            AudioServicesPlaySystemSound(SystemSoundID(kSystemSoundID_Vibrate))
            scannerDidScan(code: code)
        }
    }
```

File: Multisig/Presentation/UI Library/QRScanner/QRCodeScanner.swift [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+        
+        init(_ parent: QRCodeScanner) {
+            self.parent = parent
+        }
+    }
+}
+
+struct QRCodeScanner_Previews: PreviewProvider {
+    static var previews: some View {
+        QRCodeScanner() {_ in
```
Comment (DmitryBespalov) [inline Multisig/Presentation/UI Library/QRScanner/QRCodeScanner.swift:49]:
```suggestion
        QRCodeScanner() { _ in
```

File: Multisig/Presentation/UI Library/QRScanner/QRCodeScannerViewController.swift [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+        })
+        
+        alert.addAction(UIAlertAction(title: Strings.cameraAlertCancel, style: .default) { [weak self] _ in
+            self?.dismiss(animated: true, completion: nil)
+        })
+        present(alert, animated: true)
+    }
+
+    func presentFaildToCreateScannerAlert() {
+        let ac = UIAlertController(title: Strings.scannerNotSupportedTitle, message: Strings.scannerNotSupportedMessage, preferredStyle: .alert)
```
Comment (DmitryBespalov) [inline Multisig/Presentation/UI Library/QRScanner/QRCodeScannerViewController.swift:129]: This line is too long and needs to be split:

```suggestion
        let ac = UIAlertController(title: Strings.scannerNotSupportedTitle,
                                   message: Strings.scannerNotSupportedMessage,
                                   preferredStyle: .alert)
```

File: Multisig/Presentation/UI Library/QRScanner/QRCodeScannerViewController.swift (hunk 3 above)
Comment (DmitryBespalov) [reply Multisig/Presentation/UI Library/QRScanner/QRCodeScannerViewController.swift:82]: Also, this is a return from the `Void` method, I think the `present...()` and return should be on separate lines.

**Comment by MouazAlzahabi** [comment]
feat: QR Code Scanner Screen (#10 )

**Comment by MouazAlzahabi** [comment]
feat: QR Code Scanner Screen (#10 )

**Pull request merged**